File: day5/part1.py
def readFileAsGrid(filename):
    with open(filename, "r") as f:
        return [list(line.strip()) for line in f.readlines()]
    

def isPaperMoveable(x, y, grid):
    if not hasPaper(x, y, grid):
        return False
    
    adjacentCount = 0

    topLeft = hasPaper(x - 1, y - 1, grid)
    if topLeft:
        adjacentCount += 1
    top = hasPaper(x, y - 1, grid)
    if top:
        adjacentCount += 1
    topRight = hasPaper(x + 1, y - 1, grid)
    if topRight:
        adjacentCount += 1
    left = hasPaper(x - 1, y, grid)
    if left:
        adjacentCount += 1
    right = hasPaper(x + 1, y, grid)
    if right:
        adjacentCount += 1
    bottomLeft = hasPaper(x - 1, y + 1, grid)
    if bottomLeft:
        adjacentCount += 1
    bottom = hasPaper(x, y + 1, grid)
    if bottom:
        adjacentCount += 1
    bottomRight = hasPaper(x + 1, y + 1, grid)
    if bottomRight:
        adjacentCount += 1

    isMoveable =  adjacentCount < 4
    return isMoveable

def hasPaper(x, y, grid):
    if y < 0 or y >= len(grid):
        return False
    if x < 0 or x >= len(grid[0]):
        return False
    return grid[y][x] == "@"

def findAnswer(fileName):
    grid = readFileAsGrid(fileName)
    moveableCount = 0
    for y in range(len(grid)):
        for x in range(len(grid[0])):
            if isPaperMoveable(x, y, grid):
                print(f"Found moveable paper at ({x}, {y})")
                moveableCount += 1
    return moveableCount


testAnswer = findAnswer("test.txt")
print(f"Test answer: {testAnswer}")
assert testAnswer == 13

answer = findAnswer("input.txt")
print(f"Answer: {answer}")

File: day5/Part1.java
import java.io.*;
import java.util.*;

public class Part1 {

    static char[][] readFileAsGrid(String filename) throws IOException {
        List<String> lines = new ArrayList<>();
        try (BufferedReader br = new BufferedReader(new FileReader(filename))) {
            String line;
            while ((line = br.readLine()) != null) {
                lines.add(line.strip());
            }
        }
        char[][] grid = new char[lines.size()][];
        for (int i = 0; i < lines.size(); i++) {
            grid[i] = lines.get(i).toCharArray();
        }
        return grid;
    }

    static boolean hasPaper(int x, int y, char[][] grid) {
        if (y < 0 || y >= grid.length) return false;
        if (x < 0 || x >= grid[0].length) return false;
        return grid[y][x] == '@';
    }

    static boolean isPaperMoveable(int x, int y, char[][] grid) {
        if (!hasPaper(x, y, grid)) return false;

        int adjacentCount = 0;
        if (hasPaper(x - 1, y - 1, grid)) adjacentCount++;
        if (hasPaper(x,     y - 1, grid)) adjacentCount++;
        if (hasPaper(x + 1, y - 1, grid)) adjacentCount++;
        if (hasPaper(x - 1, y,     grid)) adjacentCount++;
        if (hasPaper(x + 1, y,     grid)) adjacentCount++;
        if (hasPaper(x - 1, y + 1, grid)) adjacentCount++;
        if (hasPaper(x,     y + 1, grid)) adjacentCount++;
        if (hasPaper(x + 1, y + 1, grid)) adjacentCount++;

        return adjacentCount < 4;
    }

    static int findAnswer(String fileName) throws IOException {
        char[][] grid = readFileAsGrid(fileName);
        int moveableCount = 0;
        for (int y = 0; y < grid.length; y++) {
            for (int x = 0; x < grid[0].length; x++) {
                if (isPaperMoveable(x, y, grid)) {
                    System.out.println("Found moveable paper at (" + x + ", " + y + ")");
                    moveableCount++;
                }
            }
        }
        return moveableCount;
    }

    public static void main(String[] args) throws IOException {
        int testAnswer = findAnswer("test.txt");
        System.out.println("Test answer: " + testAnswer);
        assert testAnswer == 13 : "Expected 13 but got " + testAnswer;

        int answer = findAnswer("input.txt");
        System.out.println("Answer: " + answer);
    }
}
